README.md

```diff
@@ -1 +1,228 @@
-# JavaScript Base-N Id Generator
+# JavaScript Base-N Id Generator _(bnid)_
+
+[![Build Status](https://travis-ci.org/digitalbazaar/bnid.png?branch=master)](https://travis-ci.org/digitalbazaar/bnid)
+
+> A JavaScript library for Web browsers and Node.js apps to generate random
+> ids and encode and decode them using various base-N encodings.
+
+## Table of Contents
+
+- [Background](#background)
+- [Install](#install)
+- [Usage](#usage)
+- [API](#api)
+- [Contribute](#contribute)
+- [Commercial Support](#commercial-support)
+- [License](#license)
+
+## Background
+
+This library provides tools for Web and Node.js to generate random ids and and
+encode and decode them in various base-N encodings. It came about from a need
+for a more compact encoding of ~128 bit ids than [UUID Version 4][] provided.
+
+The interface follows the [TextEncoder][]/[TextDecoder][] interfaces and
+provides [IdEncoder](#idencoder) and [IdDecoder](#iddecoder) classes. Instances
+can be configured with reusable encodings and other parameters. The encoder
+operates on a `Uint8Array` of input bytes. A [IdGenerator](#idgenerator) class
+is provided that can be used to generate random ids with selectable bit
+lengths, but any id data can be used.
+
+The encoder and decoder support various encodings and options:
+
+- `base16`/`hex`, `base16upper`: The simple [Base16][].
+- `base58`/`base58btc`: The [Base58][] Bitcoin alphabet as supported by
+  [base58-universal][].
+- Optional [multibase][] type prefix.
+- Fixed bit length. This is useful to ensure the output id length is constant
+  even when the id starts with an arbitrary number of zeros.
+
+### Alternatives
+
+There are
+*[many](https://www.npmjs.com/search?q=random%20id%20generator)*
+*[other](https://www.npmjs.com/search?q=id%20generator%)*
+*[random](https://www.npmjs.com/search?q=random%20id%)*
+id generators out there. Please do your research to find what best fits your
+needs.
+
+If you just want a small id generator with a customizable alphabet, start here:
+
+- [Nano ID](https://www.npmjs.com/package/nanoid)
+
+## Install
+
+### NPM
+
+```
+npm install bnid
+```
+
+### Git
+
+To install locally (for development):
+
+```
+git clone https://github.com/digitalbazaar/bnid.git
+cd bnid
+npm install
+```
+
+## Usage
+
+### Generate an ID
+
+To generate a default [Base58][], 128 bit, non-fixed-length id:
+
+```
+import {generateId} from 'bnid';
+
+const id = await generateId();
+```
+
+To generate a [Base58][], 128 bit, fixed-length id:
```

Comment:
Isn't this necessarily less than 128 bits of entropy? If so, it should be noted in the README here so people know what the trade off is for a fixed length ID.

Reply:
Why would it be less?  The fixed length feature isn't using that "set high bits" idea we had talked about.  It keeps the input random bits and zero pads the start as needed.  One tricky part about that is if you decode the value, you could end up with extra 0 bytes.  So you need to know what the size is supposed to be so you can trim decoded data.  That's wha the decode `fixedBitLength` param is for. If you're not decoding, this isn't an issue.

I'll try too add a note about this somewhere.

Reply:
Added short note. Good enough for now?

Reply:
Good enough.